### PR TITLE
Native SearchV3 UserInput

### PR DESCRIPTION
**Pull request by chromakode**
jun-02-2017 16-10-33

### CODE REVIEW

**Comment by keybase-ci-visdiff** [comment]
The commits 431936f57af6...abe739f97739 introduce visual changes on linux. <details><summary>:mag_right: 2 new, 6 changed</summary>

 * Added: **Search_V_3_user_input_editable-Users_Add_Wrap**  
   Search_V_3_user_input_editable-Users_Add_Wrap rendered
 * Added: **Search_V_3_user_input_editable-Empty**  
   Search_V_3_user_input_editable-Empty rendered
 * Changed: **Search_V_3_user_input-Users_Wrap_Add_Button** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Wrap_Add_Button-431936f57af6.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Wrap_Add_Button-abe739f97739.png)  
   Search_V_3_user_input-Users_Wrap_Add_Button comparison
 * Changed: **Search_V_3_user_input-Users_Wrap** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Wrap-431936f57af6.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Wrap-abe739f97739.png)  
   Search_V_3_user_input-Users_Wrap comparison
 * Changed: **Search_V_3_user_input-Users_Text_Wrap** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Text_Wrap-431936f57af6.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Text_Wrap-abe739f97739.png)  
   Search_V_3_user_input-Users_Text_Wrap comparison
 * Changed: **Search_V_3_user_input-Users_Text** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Text-431936f57af6.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Text-abe739f97739.png)  
   Search_V_3_user_input-Users_Text comparison
 * Changed: **Search_V_3_user_input-Users_Add** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Add-431936f57af6.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Users_Add-abe739f97739.png)  
   Search_V_3_user_input-Users_Add comparison
 * Changed: **Search_V_3_user_input-Empty_Placeholder** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Empty_Placeholder-431936f57af6.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Empty_Placeholder-abe739f97739.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496364314456-431936f57af6-abe739f97739-linux/Search_V_3_user_input-Empty_Placeholder.png)
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits 4ca73d64501a...46b846293740 introduce visual changes on linux. <details><summary>:mag_right: 3 new, 20 changed</summary>

 * Added: **Search_V_3_user_input_editable-Users_Add_Wrap**  
   Search_V_3_user_input_editable-Users_Add_Wrap rendered
 * Added: **Search_V_3_user_input_editable-Empty**  
   Search_V_3_user_input_editable-Empty rendered
 * Added: **Colors-white_0_rgba_255_255_255_0**  
   Colors-white_0_rgba_255_255_255_0 rendered
 * Changed: **Search_V_3_user_input-Users_Wrap_Add_Button** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Wrap_Add_Button-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Wrap_Add_Button-46b846293740.png)  
   Search_V_3_user_input-Users_Wrap_Add_Button comparison
 * Changed: **Search_V_3_user_input-Users_Wrap** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Wrap-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Wrap-46b846293740.png)  
   Search_V_3_user_input-Users_Wrap comparison
 * Changed: **Search_V_3_user_input-Users_Text_Wrap** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Text_Wrap-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Text_Wrap-46b846293740.png)  
   Search_V_3_user_input-Users_Text_Wrap comparison
 * Changed: **Search_V_3_user_input-Users_Text** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Text-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Text-46b846293740.png)  
   Search_V_3_user_input-Users_Text comparison
 * Changed: **Search_V_3_user_input-Users_Add** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Add-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Add-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_user_input-Users_Add.png)
 * Changed: **Search_V_3_result-Twitter_Keybase_You** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_You-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_You-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_You.png)
 * Changed: **Search_V_3_result-Twitter_Keybase_Not_Following** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_Not_Following-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_Not_Following-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_Not_Following.png)
 * Changed: **Search_V_3_result-Twitter_Keybase_Following** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_Following-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_Following-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase_Following.png)
 * Changed: **Search_V_3_result-Twitter_Keybase** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter_Keybase.png)
 * Changed: **Search_V_3_result-Twitter** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Twitter.png)
 * Changed: **Search_V_3_result-Reddit** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Reddit-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Reddit-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Reddit.png)
 * Changed: **Search_V_3_result-Keybase_No_Service_You** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_You-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_You-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_You.png)
 * Changed: **Search_V_3_result-Keybase_No_Service_Show_Tracker** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_Show_Tracker-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_Show_Tracker-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_Show_Tracker.png)
 * Changed: **Search_V_3_result-Keybase_No_Service_No_Follow** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_No_Follow-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_No_Follow-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service_No_Follow.png)
 * Changed: **Search_V_3_result-Keybase_No_Service** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_No_Service.png)
 * Changed: **Search_V_3_result-Keybase_Git_Hub_No_Fullname** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_Git_Hub_No_Fullname-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_Git_Hub_No_Fullname-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_Git_Hub_No_Fullname.png)
 * Changed: **Search_V_3_result-Keybase_Git_Hub** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_Git_Hub-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_Git_Hub-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Keybase_Git_Hub.png)
 * Changed: **Search_V_3_result-Hacker_News** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Hacker_News-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Hacker_News-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Hacker_News.png)
 * Changed: **Search_V_3_result-Git_Hub** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Git_Hub-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Git_Hub-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Git_Hub.png)
 * Changed: **Search_V_3_result-Facebook** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Facebook-4ca73d64501a.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Facebook-46b846293740.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496449648024-4ca73d64501a-46b846293740-linux/Search_V_3_result-Facebook.png)
</summary>

File: shared/dev/dumb-sheet/render.native.js [hunk 1]
```diff
@@ -262,7 +262,7 @@ class DumbSheetRender extends Component<void, Props, any> {
         <Box style={styleBox}>
           <Text type="BodySmall">{key}: {mockKey}</Text>
           <Box style={styleSmallScreen}>
-            <Box style={{flex: 1}} {...mock.parentProps}>
+            <Box {...mock.parentProps}>
```
Comment (MarcoPolo) [inline shared/dev/dumb-sheet/render.native.js]: won't this change/break every other dumb screen? Why this change?

File: shared/searchv3/dumb.js [hunk 2]
```diff
@@ -338,19 +341,60 @@ const userInputMap: DumbComponentMap<UserInput> = {
...
       userItems: chrisUsers,
       usernameText: 'Chris Hemswor',
-      showAddButton: false,
     },
   },
 }
 
+const UserInputEditable = compose(
+  withState('usernameText', 'onChangeText', ''),
+  withState('userItems', 'setUserItems', ({userItems}) => userItems),
```
Comment (MarcoPolo) [inline shared/searchv3/dumb.js]: I'm a little confused by this? You pass in a prop called userItems, that get overwritten by a state-prop also called userItems?

Comment (MarcoPolo) [inline shared/searchv3/dumb.js]: I guess it doesn't matter too much sense it's just a demo for the dumb sheet

File: shared/searchv3/user-input/index.native.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            ..._pillTextStyle,
+            color: globalColors.black_20,
+          }}
+        >
+          ,
+        </Text>
+        {/* We can't listen to keyboard events in RN Android, so instead we
+            create an offscreen text input with a single space character and
+            observe it being removed when the value changes. */}
+        {isSelected &&
```
Comment (MarcoPolo) [inline shared/searchv3/user-input/index.native.js]: neat

File: shared/searchv3/user-input/index.js.flow [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// @flow
+import {Component} from 'react'
+import * as Constants from '../../constants/searchv3'
+import type {IconType} from '../../common-adapters/icon'
+
+export type UserDetails = {
+  followingState: Constants.FollowingState,
```
Comment (MarcoPolo) [inline shared/searchv3/user-input/index.js.flow]: I added `id` to this, since a username is different than the id. e.g.: username:marcopolo, id:marcopolo@github

File: shared/searchv3/user-input/index.js.flow [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  service: Constants.Service,
+  username: string,
+}
+
+export type Props = {
+  placeholder?: string,
+  userItems: Array<UserDetails>,
+  usernameText: string,
+  onChangeText: (usernameText: string) => void,
+  onRemoveUser: (username: string) => void,
```
Comment (MarcoPolo) [inline shared/searchv3/user-input/index.js.flow]: I changed this to be `(id: string) => void`

File: shared/searchv3/user-input/index.native.js [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+type UserItemState = {isSelected: boolean, selectAnim: Animated.Value}
+
+class UserItem extends Component<void, UserItemProps, UserItemState> {
+  state = {
+    isSelected: false,
+    selectAnim: new Animated.Value(0),
+  }
+
+  _onRemoveUser = () => {
+    this.props.onRemoveUser(this.props.username)
```
Comment (MarcoPolo) [inline shared/searchv3/user-input/index.native.js]: this should pass `this.props.id`

File: shared/dev/dumb-sheet/render.native.js (hunk 1 above)
Comment (chromakode) [inline shared/dev/dumb-sheet/render.native.js]: See the commit message here https://github.com/keybase/client/pull/7244/commits/f39df51a8ca3d7d4d951a26882ea95bb808ccd3d

It doesn't seem to break other dumb screens, but it does fix this issue I ran into.

File: shared/searchv3/dumb.js (hunk 2 above)
Comment (chromakode) [inline shared/searchv3/dumb.js]: This uses the prop (provided by dumb sheet) to set the initial value of the state. It's a bit terse, but I think it's the way you do it in recompose land. I could rename it to `defaultUserItems`, if you'd prefer?

File: shared/searchv3/user-input/index.native.js (hunk 6 above)
Comment (chromakode) [inline shared/searchv3/user-input/index.native.js]: Ah gotcha, I'll update this PR w/ your id changes.

**Comment by chromakode** [comment]
Merged in https://github.com/keybase/client/pull/7283
